test/multidevice.cpp

```diff
@@ -0,0 +1,276 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#ifdef USE_DISTRIBUTED
+#include <ir/all_nodes.h>
+#include <multidevice/pipeline_ir.h>
+#include <multidevice/runtime.h>
+#include <ops/all_ops.h>
+#include <test/multidevice.h>
+#include <test/validator.h>
+#include <torch/cuda.h>
+
+namespace nvfuser {
+
+auto multidevice_env = static_cast<MultiDeviceEnvironment*>(
+    testing::AddGlobalTestEnvironment(new MultiDeviceEnvironment));
+
+void MultiDeviceEnvironment::SetUp() {
+  communicator_ = new Communicator();
+}
+
+void MultiDeviceEnvironment::TearDown() {
+  communicator_->barrier();
+  delete communicator_;
+}
+
+void MultiDeviceTest::SetUp() {
+  NVFuserTest::SetUp();
+  communicator = multidevice_env->communicator();
+  if (!communicator->is_available() || communicator->size() < 2 ||
+      torch::cuda::device_count() < 2) {
+    GTEST_SKIP() << "This test needs at least 2 GPUs and 2 ranks";
+  }
+  tensor_options =
+      at::TensorOptions().dtype(at::kFloat).device(communicator->device());
+}
+
+void CommunicationTest::SetUp() {
+  MultiDeviceTest::SetUp();
+  all_ranks = std::vector<DeviceIdxType>(communicator->size());
+  std::iota(all_ranks.begin(), all_ranks.end(), 0);
+}
+
+void CommunicationTest::validate(at::Tensor obtained, at::Tensor expected) {
+  NVF_ERROR(
+      obtained.equal(expected),
+      "Device ",
+      communicator->deviceId(),
+      " expected tensor:\n",
+      expected,
+      "\nbut obtained tensor:\n",
+      obtained);
+}
+
+void CommunicationTest::resetDstBuffers() {
```

Comment:
Why do we need to reset buffers? Does some code rely on the buffer being zero-inited? If not, could we fill `nan` instead of zeros, as it is more likely to uncover bugs?

Reply:
Makes sense. I changed to it fill the tensors with nan